File: src/org/labrad/qubits/channels/SramChannelBase.java
package org.labrad.qubits.channels;

import java.util.Map;

import org.labrad.qubits.Experiment;
import org.labrad.qubits.FpgaModelDac;
import org.labrad.qubits.resources.DacBoard;

import com.google.common.collect.Maps;

public abstract class SramChannelBase<T> implements Channel {

  String name = null;
  Experiment expt = null;
  DacBoard board = null;
  FpgaModelDac fpga = null;

  @Override
  public String getName() {
    return name;
  }

  @Override
  public Experiment getExperiment() {
    return expt;
  }

  @Override
  public void setExperiment(Experiment expt) {
    this.expt = expt;
  }

  @Override
  public DacBoard getDacBoard() {
    return board;
  }

  public void setDacBoard(DacBoard board) {
    this.board = board;
  }

  @Override
  public FpgaModelDac getFpgaModel() {
    return fpga;
  }


  //
  // Blocks
  //
  protected String currentBlock;
  public String getCurrentBlock() {
	  return currentBlock;
  }
  public void setCurrentBlock(String block) {
	  currentBlock = block;
  }

  Map<String, T> blocks = Maps.newHashMap();
}

File: QubitServer/src/main/java/org/labrad/qubits/channels/SramChannelBase.scala
package org.labrad.qubits.channels

import com.google.common.collect.Maps
import java.util.Map
import org.labrad.qubits.Experiment
import org.labrad.qubits.FpgaModelDac
import org.labrad.qubits.resources.DacBoard

abstract class SramChannelBase[T](name: String) extends SramChannel {

  private var expt: Experiment = null
  private var board: DacBoard = null
  protected var fpga: FpgaModelDac = null

  override def getName(): String = {
    name
  }

  override def getExperiment(): Experiment = {
    expt
  }

  override def setExperiment(expt: Experiment): Unit = {
    this.expt = expt
  }

  override def getDacBoard(): DacBoard = {
    board
  }

  def setDacBoard(board: DacBoard): Unit = {
    this.board = board
  }

  override def getFpgaModel(): FpgaModelDac = {
    fpga
  }


  //
  // Blocks
  //
  protected var currentBlock: String = null
  def getCurrentBlock(): String = {
    currentBlock
  }
  def setCurrentBlock(block: String): Unit = {
    currentBlock = block
  }

  protected val blocks: Map[String, T] = Maps.newHashMap()

  // Start delay
  override def getStartDelay(): Int = {
    this.getFpgaModel().getStartDelay()
  }

  override def setStartDelay(startDelay: Int): Unit = {
    this.getFpgaModel().setStartDelay(startDelay)
  }
}
